### DISCUSSION TITLE
Add prefix for block / transaction

### DISCUSSION THREAD

**Discussion by gosunuts** [aergoio/aergo#293] [General]
Trie key is saved with 's' prefix in #200. it can prevent overwritten by contract values.
But, I think that is not sufficient, because[ block](https://github.com/aergoio/aergo/blob/ba91621fbdadee47bddfb1334f4680125a8ac34b/chain/chaindb.go#L517) and [tx](https://github.com/aergoio/aergo/blob/ba91621fbdadee47bddfb1334f4680125a8ac34b/chain/chaindb.go#L489C11-L489C11) also saved without prefix. 
to prevent overwritten, block and transaction also need to add prefix when saved in db.

**Comment by kroggen** [comment]
As long as all content that is stored using hash as the key has this format below, there is no risk of conflict:

```
hash(linear content) -> content
```

It only happens with the merkle trie because the storage format is different:

```
hash(first 2 children) -> content
```

So if the hash of block and tx are computed using "linear content", there is no risk of a hash to map to a different content

**Reply by gosunuts** [reply]
it seems contract values just saved in sha256(lenear content).
below is a code that tracks how contract.LuaSetDB is stored in db.

1.  [luaSetDB](https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/contract/vm_callback.go#L91C50-L91C50)
2. [ContractState.SetData](https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/state/contract.go#L134)
3. [stateBuffer.stage](https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/state/statebuffer.go#L187)
4. [valueEntry.Hash](https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/state/statebuffer.go#L42)
5. [getHashBytes](https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/state/statebuffer.go#L213)

**Reply by gosunuts** [reply]
According to this code, when putting a block header into value in luaSetDB, value hashed to blockHash, and duplicated key saved on db.

**Reply by kroggen** [reply]
So the key prefixes used with `SetData()` are not stored on the database. The keys are hashed and the hash is stored. It is different from what I was expecting. Maybe because they are stored on a state trie?

It appears to be storing the state keys and values on the db like this:

```
hash(prefix + key) -> hash(value)

hash(value) -> value
```

So far no problem, as both hashes are computed from linear content

**Reply by kroggen** [reply]
I suspect that block headers are not stored using the `luaSetDB` function. It is probably done on the `chain` or `consensus` packages

**Reply by gosunuts** [reply]
as you said, block/tx is stored in chain package. but i concern these info can be overwrited by contract developer. ( to using LuaSetDB )

**Reply by gosunuts** [reply]
luaSetDB append prefix "_", and it`s ascii code is 95.
it means, if "linear contents" (in block, tx, etc) started to byte(95), it can be overwrited.

**Reply by gosunuts** [reply]
for example, block hash is calculated to sha256(ChainID, PrevBlockHash, BlockNo, Timestamp....), and tx hash calculated to sha256(nonce, account, recipient, amount,payload...)

I suspect chainID first value can be byte(95), but, obviously, nonce first value can be byte(95) value.

**Reply by gosunuts** [reply]
to prevent them, it can changed to prefix + hash(linear data), not hash(prefix + linear data).
( prefix + hash(prefix + linear data) is also ok. )

**Reply by kroggen** [reply]
The prefix on state trie is meant to be added to the key in the database, so it is expected to be stored as:
```
prefix + hash -> value
```

**Reply by kroggen** [reply]
~For block and txn, the hash is computed on linear content. So there is no risk of collision~

**Reply by kroggen** [reply]
I am perceiving now that the txn hash is not computed on all the bytes of the serialized transaction

If it was computed like this, there would not have any problem:

```
hash(txn_bytes) -> txn_bytes
```

But instead it is done like this:
```
hash(only some parts) -> txn_bytes
```
And the above is problematic, because it is not hashing all the serialized bytes

So a prefix is indeed required for tx hash

If the same happens with the blocks, then they need a prefix as well

**Reply by gosunuts** [reply]
https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/types/blockchain.go#L258
https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/types/blockchain.go#L524

**Reply by kroggen** [reply]
Do you know the code where the blocks and transactions are stored on the db?

**Reply by kroggen** [reply]
I did not find the "_" prefix for luaSetDB. Where is it?

**Reply by kroggen** [reply]
The schema/dbkey package united many prefixes. At the beginning I was thinking that all of them were saved on the db as prefix for the key, but now I see that some of them (those used with `SetData()`) are hashed, because they are stored inside a state trie

**Reply by kroggen** [reply]
I would be good to merge #287 so the work can be continued. If a sync test was not started yet, you could run it with tx-replay

**Reply by gosunuts** [reply]
this is set block / tx codes.

https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/chain/chaindb.go#L498
https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/chain/chaindb.go#L480

this is key prefix ( "_" ) in contract
https://github.com/aergoio/aergo/blob/47d4b320af095044ee2190a1dc4fe9f188e14ca9/contract/system_module.c#L9

**Reply by gosunuts** [reply]
I`m missing important thing. **block,tx and state are stored in the different place.** ( chain / state )
it means no worry about block / tx overwriting by state values.

**Reply by gosunuts** [reply]
by the way, it is weird about state capacity. i think chain capacity is higher than state capacity ( because many block do not have transaction ), but, actually, state db is x15 larger than chain db.

52G - chain db capacity ( in mainnet )
763G - state db capacity ( in mainnet )
24M - statesql capacity ( in mainnet )

 i`ll find reason why state db is too big.

**Reply by kroggen** [reply]
It is not capacity, it is size. State db stores a lot of hashes for the state tree. Even if a single bit is changed on some account, like increasing the nonce, the state from that account must be updated all the way up to form a new merkle tree and a new root hash. I suspect this is the reason.

**Reply by kroggen** [reply]
Indeed. I have not thought that the state data is stored on a separate database.
That is an important information. Good catch!

**Reply by kroggen** [reply]
The `STATE_DB_KEY_PREFIX` is added to the key, and then the key is hashed on the state trie. So that prefix is not stored on the database. It may be there to avoid collision with other internal keys that can be stored on the same state trie.

#200 may still be required, because the problem is not with the key but with the value. They are only converted to JSON, so strings are stored with quotes

**Reply by gosunuts** [reply]
i agree. #200 should be merged. but before that, we need to prepare for db migration.

**Reply by gosunuts** [reply]
and maybe, i can find any ideas to tidy database size. ethereum has x300 more transactions than aergo, but eth`s state size is not large. ( aergo node size is 1 TB, and ethereum full node is around 12TB )

**Reply by kroggen** [reply]
The focus of development of Aergo was speed over used disk space. It appears there is a trade off between the 2. The "trie node batches" makes it faster on read and write operations (less I/O) but uses more disk space. I don't now if there is any other thing that also affects the high disk usage.

But again, let's focus on the security issues, dependencies and the release of V4 now.
